oap-ws-sso-api/src/main/java/oap/ws/sso/interceptor/JWTSecurityInterceptor.java

```diff
@@ -29,52 +29,68 @@
+    private final UserProvider userProvider;
 
-    public JWTSecurityInterceptor( AbstractTokenProvider tokenProvider ) {
+    public JWTSecurityInterceptor( AbstractTokenProvider tokenProvider, UserProvider userProvider ) {
         this.tokenProvider = tokenProvider;
+        this.userProvider = userProvider;
     }
 
     @Override
     public Optional<Response> before( InvocationContext context ) {
-        final String authorization = context.exchange.getRequestHeader( AUTHENTICATION_KEY );
-        if( authorization == null || authorization.isEmpty() ) {
-            log.trace( "Not authenticated. No token in header {}", context.exchange );
-        }
-        final String token = extractBearerToken( authorization );
-        if( !tokenProvider.verifyToken( token ) ) {
-            log.trace( "Not authenticated." );
-            return Optional.of( new Response( FORBIDDEN, "Invalid token" ) );
+        var jwtToken = SSO.getAuthentication( context.exchange );
+        if( jwtToken == null )
+            log.trace( "Not authenticated" );
+        else {
+            final String token = extractBearerToken( jwtToken );
+            if( token == null || !tokenProvider.verifyToken( token ) ) {
+                log.trace( "Not authenticated." );
+                return Optional.of( new Response( FORBIDDEN, "Invalid token" ) );
+            }
+
+            final String email = tokenProvider.getUserEmail( token );
+            User user = userProvider.getUser( email ).orElse( null );
+            if( user != null ) {
+                context.session.set( SESSION_USER_KEY, user );
+                log.trace( "set user {} into session {}", user, context.session );
+            } else
+                log.trace( "User not found with email: " + email );
         }
 
+        final List<String> permissions =
+            jwtToken != null ? tokenProvider.getPermissions( extractBearerToken( jwtToken ) ) : null;
         Optional<WsSecurity> wss = context.method.findAnnotation( WsSecurity.class );
         if( wss.isEmpty() )
             return Optional.empty();
 
-        log.trace( "secure method {}", context.method );
+        log.trace( "Secure method {}", context.method );
 
         Optional<String> realm =
             SYSTEM.equals( wss.get().realm() ) ? Optional.of( SYSTEM ) : context.getParameter( wss.get().realm() );
         if( realm.isEmpty() )
             return Optional.of( new Response( FORBIDDEN, "realm is not passed" ) );
-
-        final List<String> permissions = tokenProvider.getPermissions( token );
-        if( Arrays.stream( wss.get().permissions() ).anyMatch( permissions::contains ) )
-            return Optional.empty();
+        if( permissions != null ) {
+            if( Arrays.stream( wss.get().permissions() ).anyMatch( permissions::contains ) )
+                return Optional.empty();
+        }
         return Optional.of( new Response( FORBIDDEN, "user doesn't have permissions" ) );
```

Comment:
could you add which permission is required?

Reply:
added